### PR TITLE
planner: refactor `tryToGetIndexJoin`

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #45520

Problem Summary: planner: refactor `tryToGetIndexJoin`

### What is changed and how it works?

planner: refactor `tryToGetIndexJoin`

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -1993,61 +1992,16 @@ func (p *LogicalJoin) satisfyIndexJoinHint(join PhysicalPlan) bool {
 	case *PhysicalIndexMergeJoin:
 		innerIdx = ij.getInnerChildIdx()
 		joinMethod = indexMergeJoin
-	default:
-		return false
```
Comment (hawkingrei) [inline planner/core/exhaust_physical_plans.go:1997]: Maybe add ```panic("unreachable")```

Comment (qw4990) [reply planner/core/exhaust_physical_plans.go:1997]: I add a new returned value here to indicate whether this plan is a IndexJoin.

**Review by chrysan — APPROVED** [review]

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -2059,47 +2016,78 @@ func (p *LogicalJoin) tryToGetIndexJoin(prop *property.PhysicalProperty) (indexJ
...
+
+// handleForceIndexJoinHints handles the force index join hints and returns all plans that can satisfy the hints.
+func (p *LogicalJoin) handleForceIndexJoinHints(prop *property.PhysicalProperty, candidates []PhysicalPlan) (indexJoins []PhysicalPlan, canForced bool) {
+	if !p.preferAny(preferRightAsINLJInner, preferRightAsINLHJInner, preferRightAsINLMJInner,
+		preferLeftAsINLJInner, preferLeftAsINLHJInner, preferLeftAsINLMJInner) {
+		return candidates, false // no force index join hints
+	}
+
+	const left, right = 0, 1
+	const indexJoin, indexHashJoin, indexMergeJoin = 0, 1, 2
```
Comment (time-and-fate) [inline planner/core/exhaust_physical_plans.go:2044]: We defined the same thing in `getIndexJoinSideAndMethod`. It's better to combine them into one definition.

**Review by time-and-fate — APPROVED** [review]

File: planner/core/exhaust_physical_plans.go (hunk 2 above)
Comment (qw4990) [reply planner/core/exhaust_physical_plans.go]: updated, PTAL

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**